ramips: add support for cudy WR2100

yogo1212 opened the pull request:
This is the current state of my effort for creating an openwrt image for the [cudy WR2100](http://www.cudy.com/productinfo/105182.html).

From my end, there are two big open points:
- the image does not boot (kindly tested by cudy devs)
- the openwrt installation procedure must work around signature validation

For details, see the current commit description.
There are also a few TODOs in the device tree file and there's probably something missing in `/etc/boards.d`.

yogo1212 committed:
ramips: add support for cudy WR2100

Specifications

  SoC:       MT7621
  CPU:       880 MHz
  Flash:     16 MiB
  RAM:       128 MiB
  WLAN:      2.4 GHz b/g/n, 5 GHz a/n/ac
             MT7603E / MT7615E
  Ethernet:  5x Gbit ports

Installation

There are two known options:
1) The Luci-based UI.
2) Press and hold the reset button during power up.
   The router will request 'recovery.bin' from a TFTP server at
   192.168.1.88.

Both options require a signed firmware binary.
The openwrt image supplied by cudy is signed and can be used to
install unsigned images.

R4 & R5 need to be shorted (0-100Ω) for the UART to work.

Signed-off-by: Leon M. George <leon@georgemail.eu>

adschm commented:
Typo in the commit title: "ramips"
Apart from that, "Cudy" should have capital letter?

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+		compatible = "gpio-leds";
+
+		led_blue: blue {
+			label = "green:blue";
```
Honestly?

yogo1212 replied:
i shit you not: these are the labels from the stock device tree based on MTK. i only removed the `oem:` at the beginning.
so far, there's only been blue and red - so i assume, the green is garbage.
once there's something bootable, i'll confirm the colours.

yogo1212 replied:
```
echo 255 > oem\:green\:red/brightness
echo 255 > oem\:green\:blue/brightness
```
they turn on a red and a blue LED. the only greens are the ethernet ports.
are `red` and `blue` sufficing labels?

adschm replied:
well, just because the stock device tree is nonsense, I don't want to transfer that into our project. Just properly name these with color:function due to what they are supposed to do

yogo1212 replied:
at no point did i say we should do that.

i will ask cudy about how they intended the LEDs to be used.
as of today, i don't know whether the red/blue LEDs have a specific purpose.
atm, i can only say that, in the stock firmware, `sysupgrade` cycles through >all< LEDs (LAN1,2,3,4,red,blue,WAN), the blue LED is active without a recognizable pattern, and the red LED hardly ever goes on.

yogo1212 replied:
let's get the wifi working first.
without wifi, there's hardly any use talking about the intended purpose of LEDs.

yogo1212 replied:
@adschm hey :-)
i played around with the stock firmware and, afaics, the blue+red LEDs are used in combination to reflect the detected uplink status.

red => no internet, blue -> internet

yogo1212 replied:
i will name the LEDs `uplink`. is there a more 'openwrt' way of doing this? do you prefer another name?

adschm replied:
Well, I'd have thought of red:internet and blue:internet, but it's essentially a matter of taste.

yogo1212 replied:
i've gone with your suggestion. resolve this conversation?

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+		};
+
+		red {
+			label = "green:red";
```
See above.

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>; // TODO increase to 100?
```
Please use gpio-keys instead and drop poll-interval.

yogo1212 replied:
will do!

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
```
Memory is autodetected, node can be dropped.

yogo1212 replied:
cool :+1:

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can this go faster?

yogo1212 replied:
i will test once my image boots properly.

yogo1212 replied:
`50000000` survived a few sysupgrades.

yogo1212 replied:
should i test any higher values?

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+		m25p,chunked-io = <32>;
```
https://github.com/openwrt/openwrt/commit/a66449c60893a168fa0c13a29127d29314b2e348

yogo1212 replied:
done

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+		partitions {
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
```
read-only ...
Also below ...

yogo1212 replied:
done

adschm commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,241 @@
+				read-only;
+			};
+
+			debug: partition@fd0000 {
```
Only add DT labels where needed.

yogo1212 replied:
removed `debug` and `backup` labels

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
 
+define Device/cudy-wr2100
+  $(Device/dsa-migration)
+  DEVICE_VENDOR := cudy
```
Cudy?

yogo1212 replied:
ok. capitalised `DEVICE_VENDOR`

yogo1212 commented:
@adschm 
> Typo in the commit title: "ramips"
fixed

> Apart from that, "Cudy" should have capital letter?
The manufacturer uses lower case letters on the box and on the device itself (and on their website `cudy.com`).
I've noticed, that the `DEVICE_VENDOR` string is capitalised generally.

For the WR1000 (already merged), only `DEVICE_VENDOR` and `model` (DTS) are capitalised.
If we use a different capitalisation here, we should probably change the strings there as well.

yogo1212 commented:
We need to talk about the installation procedure.
Is there an openwrt-way of handling code signing?

The router is advertised as an "openwrt support device" and, so far, cudy has been very friendly and helpful.
I think they are open for suggestions.

It might even be possible to have them sign an "openwrt installation firmware". But what might that look like?
All ports in a bridge and then find a TFTP server via DHCP? Users would have to set up fancy DHCP options and such.

I'm currently trying to get a u-boot prompt to get around all of that (not the most convenient installation method, if it works at all).

yogo1212 commented:
status update - i can't get past this

```
[    0.589656] spi-mt7621 1e000b00.spi: sys_freq: 220000000<0d>
[    0.605152] spi-nor spi0.0: w25q128 (16384 Kbytes)<0d>
...
[    0.675203] mt7621-pci 1e140000.pcie: Parsing DT failed<0d>
...
[    1.359945] /dev/root: Can't open blockdev<0d>
[    1.364093] VFS: Cannot open root device "(null)" or unknown-block(0,0): error -6<0d>
[    1.371556] Please append a correct "root=" boot option; here are the available partitions:<0d>
[    1.379895] 1f00           16384 mtdblock0 <0d>
[    1.379899]  (driver?)<0d>
[    1.386430] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)<0d>
[    1.394690] Rebooting in 1 seconds..<0d>
```

musashino205 commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,236 @@
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
```
add the following properties after here:

```
compatible = "fixed-partitions";
#address-cells = <1>;
#size-cells = <1>;
```

yogo1212 replied:
@musashino205 thank you! :heart:

musashino205 commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,244 @@
+	status = "okay";
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
```
```#address-cells``` and ```#size-cells``` here have no effect, they can be dropped.

yogo1212 replied:
good to know - will do!

musashino205 commented on target/linux/ramips/dts/mt7621_cudy-wr2100.dts:
```diff
@@ -0,0 +1,244 @@
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		//device_type = "pci";
```
```device-type = "pci"``` is specified in the root of pcie in *mt7621.dtsi* and it's a property for PCI host controller. So this can be dropped. (same for line 127)

yogo1212 replied:
i assumed as much.
but because there are still difficulties with the wifi adapters, i kept those lines around.
with the confirmation, that those lines don't have any effect, i'm dropping them

yogo1212 commented:
current fields of work:
- ethernet ports only work when the corresponding LED is turned on manually (ethernet activity doesn't turn on the LED, instead the LED turns on the ethernet link)
- no wifi devices in /sys/class/ieee80211/

musashino205 commented:
> no wifi devices in /sys/class/ieee80211/

Drivers for wifi devices are missing. For MT7603 and MT7615E need to be added to ```DEVICE_PACKAGES``` in the device definition in image/mt7621.mk.

- for MT7603: ```kmod-mt7603```
- for MT7615: ```kmod-mt7615e``` and ```kmod-mt7615-firmware```

yogo1212 commented:
don't know how: ethernet problem has gone away.
maybe i was chasing a ghost.

fixing the sysupgrade platform check and validating the LED behaviour atm.

yogo1212 commented:
the port LEDs are link/traffic indicators in stock. those should work with the last push.
`cudy-wr2100` was renamed to `cudy_wr2100`.

The stock `board_name` is `R11`. That shouldn't be added to `SUPPORTED_DEVICES`, right?

stock uses both gmacs to connect to the switch:
```
VLAN 1:
        vid: 1
        ports: 0 1 2 3 6 
VLAN 2:
        vid: 2
        ports: 4 5
```
this is unfit for openwrt, right?
i've kept the definitions for `gmac1` and `port@5` in comments. should i delete them?

yogo1212 commented:
the ethernet problem is there again :-(
ethernet link doesn't come up unless the LED is turned on.
what might cause that?!
i don't suppose the LEDs are connected to the switch electrically.

adschm commented:
> The stock board_name is R11. That shouldn't be added to SUPPORTED_DEVICES, right?

Only if it's really required for upgrading from factory.

yogo1212 commented:
and the ethernet problem is gone again :-|
don't know what that is.
it's been working for a few builds now.

> [adding R11 to supported devices]
> Only if it's really required for upgrading from factory.

In the meantime, i found this:
Even with `R11` in `SUPPORTED_DEVICES`, sysupgrade doesn't work with `-F`.
This is how the file is checked:
```
root@WR2100:~# oem-check -b "R11" -R -o 0x30000 -f /tmp/openwrt-ramips-mt7621-cudy_wr2100-squashfs-sysupgrade.bin 
lede: firmware format error
```

I've tried various combinations for `-b` - including the 'correct' openwrt versions. None work.
`oem-check` is a binary and i can't find where it comes or how it works.

so, i would say R11 is not needed since the detection doesn't work anyway.

yogo1212 commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,229 @@
+	phy-mode = "trgmii";
+};
+
+// &gmac1 {
```
stock uses both gmacs to connect to the switch:
```
VLAN 1:
        vid: 1
        ports: 0 1 2 3 6 
VLAN 2:
        vid: 2
        ports: 4 5
```
this is unfit for openwrt, right?
i've kept the definitions for `gmac1` and `port@5` in comments. should i delete them?

adschm replied:
Either you need them or you remove them.

yogo1212 replied:
i'm removing it, then.

for future reference and too allow people to make use of the additional link more easily, here are the removed nodes:

```
&gmac1 {
	status = "okay";

	mtd-mac-address = <&bdinfo 0xde00>;
	phy-mode = "rgmii"; // TODO overrides rgmii-rxid

	fixed-link {
		speed = <1000>;
		full-duplex;
		pause;
	};
};

&switch0 {
	ports {
		port@5 {
		 reg = <5>;
		 label = "cpu2";

		 ethernet = <&gmac1>;
		 phy-mode = "rgmii";

		 fixed-link {
			 speed = <1000>;
			 full-duplex;
		 };
		};
	}
}
```

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,229 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_blue: blue {
```
led_internet_blue: internet_blue {

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,229 @@
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+		};
+
+		red {
```
internet_red

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,229 @@
+	keys {
+		compatible = "gpio-keys";
+		#address-cells = <1>;
+		#size-cells = <0>;
```
Don't think the *-cells are needed here.

yogo1212 replied:
you're right

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
+  $(Device/dsa-migration)
+  DEVICE_VENDOR := Cudy
+  DEVICE_MODEL := WR2100
+  IMAGE_SIZE := 16064k
```
That's not 0xf80000 ...

yogo1212 replied:
oh :deer: 

thanks for pointing it out!

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -12,6 +12,13 @@ asus,rt-n56u-b1)
+	ucidef_set_led_netdev "lan2" "lan2 link" "green:lan2" "lan2"
+	ucidef_set_led_netdev "lan3" "lan3 link" "green:lan3" "lan3"
+	ucidef_set_led_netdev "lan4" "lan4 link" "green:lan4" "lan4"
+	ucidef_set_led_netdev "wan" "wan link" "green:wan" "wan"
```
Adding link to the name seems strange here, as this will also show activity?

I'd just call it lan1, lan2, etc. ?

yogo1212 replied:
oops! yeah.
sloppy copy mistake.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -82,6 +82,11 @@ ramips_setup_macs()
 		label_mac=$(mtd_get_mac_binary factory 0x4)
 		;;
+	cudy,wr2100)
+		lan_mac=$(mtd_get_mac_binary bdinfo 0xde00) # "factory 0x4" also does it
```
This is already set in DTS with gmac0 and can be removed.

yogo1212 replied:
done

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -82,6 +82,11 @@ ramips_setup_macs()
 		;;
+	cudy,wr2100)
+		lan_mac=$(mtd_get_mac_binary bdinfo 0xde00) # "factory 0x4" also does it
+		wan_mac=$(macaddr_add "$lan_mac" 1)
```
Please set this in DTS via the wan port (just look at other devices.)

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -82,6 +82,11 @@ ramips_setup_macs()
+	cudy,wr2100)
+		lan_mac=$(mtd_get_mac_binary bdinfo 0xde00) # "factory 0x4" also does it
+		wan_mac=$(macaddr_add "$lan_mac" 1)
+		label_mac=$lan_mac
```
This can be set in DTS via label-mac-device in aliases.

yogo1212 commented:
i've added a commit that removes trailing whitespaces in two instances that you can do with as you please.

yogo1212 commented:
@adschm 
thank you for the feedback and the corrections!

and thanks to @musashino205 as well for his prompt help :-)

i'm removing the WIP from the title as i'm now happy with the state of the port.

adschm commented:
Whitespace patch has been merged.

yogo1212 commented:
@adschm cudy has notified me that the `R11` check works through the UIMAGE_NAME.
i've added a commit to set it.

with it, it is possible to upgrade from cudy openwrt without `-F`.
the image can then also be used for direct upgrades from stock or recovery (after being signed by cudy).

vale-max commented:
I tried to build and flash an image generated from this repo. It works, the device boots and connects and everything as expected but wireless performance are disappointing.
Cudy provides an image based on 19.07 snapshot, with that one I get up to 500mbps of transfer speed while with the "master" build sometimes reaches 280mbps and most of the times stays around 100mbps, with same configuration of channels, bandwith and options.

I also built an image based on latest openwrt-19.07 branch commit applying the files that cudy provides in his "openwrt for developers" archive, it works very well, exactly as cudy's image.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,204 @@
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0xf80000>;
+			};
```
Please move that to the correct position.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,204 @@
+			partition@fe0000 {
+				label = "backup";
+				reg = <0xfe0000 0x10000>;
+			};
```
This looks like it should be read-only?

yogo1212 replied:
@adschm it appears backup and debug contain at least parts of the overlay.
the stock firmware writes to those sections when running sysupgrade.

i will make both read-only.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr2100.dts:
```diff
@@ -0,0 +1,204 @@
+
+&state_default {
+	gpio {
+		groups = "jtag", "wdt";
```
The seem to be several groups missing here ...

yogo1212 replied:
in the cudy device tree, `groups` contains wdt, rgmii2, jtag, and mdio.
that device tree is for the MTK build.
the openwrt device tree (M2100) only includes jtag and wdt.

i've tried a few times to add rgmii2 and mdio and, if i remember correctly, the images didn't work properly (kernel panic, no ethernet).
frankly, i don't understand why that should be but at some point i've given up trying to fix something that appears to work.

i will try again tomorrow.

adschm replied:
Well, from your GPIOs, the correct groups should be
i2c, uart3, uart2, jtag, wdt

Don't know why cudy device tree contains something different, if that's really the case ...

yogo1212 replied:
dude, why would i be lying? [1]

so, you also see no particular reason to set rgmii2 and mdio pins in GPIO mode.

and
- I2C (SCLK, SD)
- JTAG (TDI, TCLK, TMS, TDO)
- UART2 (CTS_N)
- UART3 (RXD)
- WDT (RST_N)
need to be in GPIO mode.

thanks!
i was really unsure about that.

:heart: 

[1] http://www.cudy.com/openwrt_software_download

yogo1212 commented:
@vale-max hi :-)
and thank you for your feedback!

i've had varying results for the wifi speed.

there are a few more people interested in a port for 19.07 and so, i keep a version around [here](https://github.com/yogo1212/openwrt/tree/cudy_wr2100_19.07).

[this](https://github.com/openwrt/openwrt/files/5895988/diff_cudy.txt) is how the cudy files compare against it.

yogo1212 commented:
@adschm in https://github.com/openwrt/openwrt/commit/c7842ceaaa27a72146303362e24e10fbb1bcb213, you replaced a `pci14c3,7662` with `mediatek,mt76` just like i did here.
was the device still detected as 7665 or as 7615 on boot?

i'm getting this with `mediatek,mt76`:
```
[    1.280959] pci 0000:02:00.0: [14c3:7615] type 00 class 0x000280
```

vale-max commented:
> i've had varying results for the wifi speed (but never really bothered much).

The wifi speed (and the HW NAT offload) is actually the reason why this cheap little thing took place of my Archer C7

> (EDIT: pcie1 is detected as `pci14c3,7615` instead of `pci14c3,7662` - damn)

Isn't pci14c3,7615 the correct one for this wifi chipset? 7662 is a wifi/bluetooth chip.

The DTS I used to compile the firmware I'm using now has:

`&pcie1 {
        wifi@1,0 {
                compatible = "pci14c3,7662";
                reg = <0x0000 0 0 0 0>;
                mediatek,mtd-eeprom = <&factory 0x8000>;
                ieee80211-freq-limit = <5000000 6000000>;
        };
`
however the bootlog says:
`[    1.988819] pci 0000:02:00.0: [14c3:7615] type 00 class 0x000280
[    1.988869] pci 0000:02:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]
`
and the device itself on /sys:
`root@Cudy2100:/sys/devices/pci0000:00/0000:00:01.0/0000:02:00.0# cat uevent 
DRIVER=mt7615e
OF_NAME=wifi
OF_FULLNAME=/pcie@1e140000/pcie@1,0/wifi@1,0
OF_COMPATIBLE_0=pci14c3,7662
OF_COMPATIBLE_N=1
PCI_CLASS=0280
PCI_ID=14C3:7615
PCI_SUBSYS_ID=7615:14C3
PCI_SLOT_NAME=0000:02:00.0
MODALIAS=pci:v000014C3d00007615sv00007615sd000014C3bc00sc02i80
`
All this with 19.07 snapshot, whose wifi speed is really good.
I didn't take note how it is detected with latest master branch but was slowish.

yogo1212 commented:
@vale-max 

> Isn't pci14c3,7615 the correct one for this wifi chipset? 7662 is a wifi/bluetooth chip.
Ok. So, the MTK device tree is misleading but it does work regardless.

> I didn't take note how it is detected with latest master branch but was slowish.
It also detects the 7615. It was a guess. No idea what might be causing the slow transmission rates.
I can't really test the speed. There are always ~10 SSIDs around me..

pyckle commented:
Hey @yogo1212 - thanks for your work on this. When you get a chance, could you please rebase from master? My changes to add the WR1300 which were merged caused a conflict in the mt7621 makefile - they should be easy to resolve - I think your changes simply go after mine.

yogo1212 commented:
@pyckle sure - consider it done!

pyckle commented:
@yogo1212 Thanks (:

yogo1212 commented:
rebased, moved the `leds` beneath the `keys` section for alphabetic ordering, and updated the commit description to mention the signed openwrt image supplied by cudy.

yogo1212 commented:
now testing the speed issue

yogo1212 commented:
possibly related issue: https://github.com/openwrt/mt76/issues/493

can't confirm, though, because in the latest trunc-image, wifi appears to be wonky in general.

```
Your IP Address
192.168.0.124:42672

Download
5.32 Mbit/s

Upload
5.71 Mbit/s

Latency
2.17 ms, 0.67 ms jitter

powered by LibreSpeed
```

no traffic through batman-mesh on 802.11s - only control frames. it was working with the image from january.

several images for ath79 and ipq40xx from the same commit don't have that problem, so i believe it has something to do with mt76 - unless, of course, i'm making a horrible mistake somewhere..

zachwelch commented:
I pulled down the v19.07 branches from @yogo1212's and @vale-max's repos. As @yogo1212 has been driving this PR, I re-based branch his `cudy_wr2100_19.07` branch to the top the current `openwrt-19.07` branch, then I rebased @vale-max's `v19.07.07-cudy` branch on top of that branch. This revealed the small set of differences between the two branches.

I went through the hunks in that patch one-by-one to merge the best from both, giving preference to the version that was previously reviewed here.  I kept @yogo1212's LED node naming, partition map, and pinctrl gpio groups, and I kept @vale-max's more precise wlan LED label, flash node name, pcie compatible strings, and dropping the `80211ieee-freq-limit` property for `pcie0`'s `wifi` node. I'm not asserting that these choices were correct; they were my best guesses at the time.

In the spirit of helping these changes get into mainline, I pushed all three rebased branches to my clone at https://github.com/mandolincreek/openwrt. Other than GitHub's requirement for a second reviewer, what needs to be done to get this merged? Collaboration on performance improvements will be easier after the baseline implementation gets merged into upstream branch.

yogo1212 commented:
@zachwelch i also maintain a backport [here](https://github.com/yogo1212/openwrt/tree/cudy_wr2100_19.07) and i will try get it merged once this reaches master.

concerning device backports: [wiki](https://openwrt.org/docs/guide-developer/device-support-policies#device_support_backports)
i wouldn't rely on the backport for openwrt-19.07 being merged ;-)

this PR is stuck with a speed issue which i believe has something to do with the mt76 driver.
a colleague of mine is digging through the code. i'm only doing this in my free time and not putting in much work atm.
in the backport, the speed appears to be closer to  cudy's openwrt release.

vale-max commented:
Cudy official openwrt is based on snapshot, the kernel magic number is unique so several modules can't be added. Wireguard seems to be the most commonly used that can't be added.
Now they included wireguard in the official firmware, but I, for example, use also fwknopd which is not built in, so I need to build my image  with both wireguard and fwknopd.

ksuuk commented:
@vale-max 

But there are yogo1212, vale-max and ztw branches, which one is suitable to use?

BTW, please fix Your repo, as once I tried to build it, it had libcap missing, as according to the https://git.openwrt.org/feed/packages.git:

2021-04-05 Stijn Tintel	libcap: drop from feed
libcap: drop from feed
The libcap package was moved to OpenWrt base."

Also:
feeds/packages/net/wg-installer/Makefile
feeds/packages/utils/attendedsysupgrade-common/Makefile
are  missing the VERSION fields.

yogo1212 commented:
@ksuuk
people building their openwrt from my branch wasn't intended but that's the way it is..

your package repo versions don't match the openwrt version.
the fix is is either a rebase on trunc, a version freeze for your clones the packages repos through feeds.conf (`;$commit`), or just a manual checkout of an appropriate commit.

i've pushed a rebase of my branch to make it easier for people to use.

adschm commented:
This looks okay now. However, I have no idea about trgmii mode, so somebody should have a final look whether setup is correct.

frankis5 commented:
For those that have this device working with Openwrt (either manufactures build on one of these branches), can you tell me what is the maximum power available for each of the 2.4G and 5G wireless in settings?    Thanks.

yogo1212 commented:
@frankis5 is this what you're looking for?
(country: DE)
```
$ iw phy
                        * 5180 MHz [36] (23.0 dBm)
                        * 5200 MHz [40] (23.0 dBm)
                        * 5220 MHz [44] (23.0 dBm)
                        * 5240 MHz [48] (23.0 dBm)
                        * 5260 MHz [52] (20.0 dBm) (radar detection)
                        * 5280 MHz [56] (20.0 dBm) (radar detection)
                        * 5300 MHz [60] (20.0 dBm) (radar detection)
                        * 5320 MHz [64] (20.0 dBm) (radar detection)
                        * 5500 MHz [100] (24.0 dBm) (radar detection)
                        * 5520 MHz [104] (24.0 dBm) (radar detection)
                        * 5540 MHz [108] (24.0 dBm) (radar detection)
                        * 5560 MHz [112] (24.0 dBm) (radar detection)
                        * 5580 MHz [116] (24.0 dBm) (radar detection)
                        * 5600 MHz [120] (24.0 dBm) (radar detection)
                        * 5620 MHz [124] (24.0 dBm) (radar detection)
                        * 5640 MHz [128] (24.0 dBm) (radar detection)
                        * 5660 MHz [132] (24.0 dBm) (radar detection)
                        * 5680 MHz [136] (24.0 dBm) (radar detection)
                        * 5700 MHz [140] (24.0 dBm) (radar detection)
                        * 5720 MHz [144] (13.0 dBm) (radar detection)
                        * 5745 MHz [149] (13.0 dBm)
                        * 5765 MHz [153] (13.0 dBm)
                        * 5785 MHz [157] (13.0 dBm)
                        * 5805 MHz [161] (13.0 dBm)
                        * 5825 MHz [165] (13.0 dBm)
                        * 5845 MHz [169] (13.0 dBm)
                        * 5865 MHz [173] (13.0 dBm)
...
                        * 2412 MHz [1] (20.0 dBm)
                        * 2417 MHz [2] (20.0 dBm)
                        * 2422 MHz [3] (20.0 dBm)
                        * 2427 MHz [4] (20.0 dBm)
                        * 2432 MHz [5] (20.0 dBm)
                        * 2437 MHz [6] (20.0 dBm)
                        * 2442 MHz [7] (20.0 dBm)
                        * 2447 MHz [8] (20.0 dBm)
                        * 2452 MHz [9] (20.0 dBm)
                        * 2457 MHz [10] (20.0 dBm)
                        * 2462 MHz [11] (20.0 dBm)
                        * 2467 MHz [12] (20.0 dBm)
                        * 2472 MHz [13] (20.0 dBm)
                        * 2484 MHz [14] (disabled)
 ```

frankis5 commented:
> 
> 
> @frankis5 is this what you're looking for?
> (country: DE)
> 
> ```
> $ iw phy
>                         * 5180 MHz [36] (23.0 dBm)
>                         * 5200 MHz [40] (23.0 dBm)
>                         * 5220 MHz [44] (23.0 dBm)
>                         * 5240 MHz [48] (23.0 dBm)
>                         * 5260 MHz [52] (20.0 dBm) (radar detection)
>                         * 5280 MHz [56] (20.0 dBm) (radar detection)
>                         * 5300 MHz [60] (20.0 dBm) (radar detection)
>                         * 5320 MHz [64] (20.0 dBm) (radar detection)
>                         * 5500 MHz [100] (24.0 dBm) (radar detection)
>                         * 5520 MHz [104] (24.0 dBm) (radar detection)
>                         * 5540 MHz [108] (24.0 dBm) (radar detection)
>                         * 5560 MHz [112] (24.0 dBm) (radar detection)
>                         * 5580 MHz [116] (24.0 dBm) (radar detection)
>                         * 5600 MHz [120] (24.0 dBm) (radar detection)
>                         * 5620 MHz [124] (24.0 dBm) (radar detection)
>                         * 5640 MHz [128] (24.0 dBm) (radar detection)
>                         * 5660 MHz [132] (24.0 dBm) (radar detection)
>                         * 5680 MHz [136] (24.0 dBm) (radar detection)
>                         * 5700 MHz [140] (24.0 dBm) (radar detection)
>                         * 5720 MHz [144] (13.0 dBm) (radar detection)
>                         * 5745 MHz [149] (13.0 dBm)
>                         * 5765 MHz [153] (13.0 dBm)
>                         * 5785 MHz [157] (13.0 dBm)
>                         * 5805 MHz [161] (13.0 dBm)
>                         * 5825 MHz [165] (13.0 dBm)
>                         * 5845 MHz [169] (13.0 dBm)
>                         * 5865 MHz [173] (13.0 dBm)
> ...
>                         * 2412 MHz [1] (20.0 dBm)
>                         * 2417 MHz [2] (20.0 dBm)
>                         * 2422 MHz [3] (20.0 dBm)
>                         * 2427 MHz [4] (20.0 dBm)
>                         * 2432 MHz [5] (20.0 dBm)
>                         * 2437 MHz [6] (20.0 dBm)
>                         * 2442 MHz [7] (20.0 dBm)
>                         * 2447 MHz [8] (20.0 dBm)
>                         * 2452 MHz [9] (20.0 dBm)
>                         * 2457 MHz [10] (20.0 dBm)
>                         * 2462 MHz [11] (20.0 dBm)
>                         * 2467 MHz [12] (20.0 dBm)
>                         * 2472 MHz [13] (20.0 dBm)
>                         * 2484 MHz [14] (disabled)
> ```

@ yogo1212  Yes, thank you for the info.

myBinn commented:
https://github.com/zxg120/openwrt/tree/openwrt-21.02

Maybe you guys should see this, this guy seems make a set of codes to support cudy router, not only WR2100, even WiFi 6 router X6...

yogo1212 commented:
@myBinn thx for the info!
i was planning on doing the x6 after the wr2100.
but if @zxg120 is interested in merging, they should have the honor.

<details><summary>diff</summary>
<p>
```
< zxg120
> yogo1212
...
  &gmac0 {
  	mtd-mac-address = <&bdinfo 0xde00>;
> 	phy-mode = "trgmii";
  };
...
  &pcie0 {
  	wifi@0,0 {
< 		compatible = "pci14c3,7603";
> 		compatible = "mediatek,mt76";
  		reg = <0x0000 0 0 0 0>;
  		mediatek,mtd-eeprom = <&factory 0x0000>;
< 		mtd-mac-address = <&bdinfo 0xde00>;
  		ieee80211-freq-limit = <2400000 2500000>;
< 
< 		led {
< 			led-active-low;
< 		};
  	};
  };
...
  &pcie1 {
  	wifi@0,0 {
< 		compatible = "pci14c3,7615";
> 		compatible = "mediatek,mt76";
  		reg = <0x0000 0 0 0 0>;
  		mediatek,mtd-eeprom = <&factory 0x8000>;
< 		mtd-mac-address = <&bdinfo 0xde00>;
< 		mtd-mac-address-increment = <2>;
  		ieee80211-freq-limit = <5000000 6000000>;
< 
< 		led {
< 			led-sources = <2>;
< 			led-active-low;
< 		};
  	};
  };
...
> 		port@6 {
> 			phy-mode = "trgmii";
> 		};
```
</p>
</details>

zxg120's differences:
no `trgmii`. i included it to be closer to cudy's dts but it works without (defaults to rgmii, i think).
@adschm just drop it?

both radios use the same mac adress as gmac0. for 5ghz, `mtd-mac-address-increment` is 2. 
there are also `led` nodes inside the wifi definitions but there aren't any wifi LEDs on the router.

blocktrron commented:
Merged with changes to my staging tree, thanks!
